Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -1301,6 +1301,11 @@ static ErrorOr<void> read_start_of_frame(JPEGStream& stream, JPEGLoadingContext&
         component.sampling_factors.horizontal = subsample_factors >> 4;
         component.sampling_factors.vertical = subsample_factors & 0x0F;
 
+        if (component_count == 1) {
+            // 4.8.2 Minimum coded unit: "If the compressed image data is non-interleaved, the MCU is defined to be one data unit."
+            component.sampling_factors = { 1, 1 };
```

Comment:
If I understand correctly, it's not that we did something wrong, but that the original encoder made a non-compliant image.
This look like a "let's follow the spec" comment, it's not explicit that we are correcting a faulty image on the fly. A small comment on that would be nice.

Reply:
No, I think the original image is valid. It's weird to have 2x2 sampling factors on a 1-channel image, but it's permitted. And the spec says to treat non-interleaved streams has having 1x1 MCUs, even if they have a sampling factor set.

(We still get this wrong if there are sampling factors set and it's a multi-channel non-interleaved image I think. And that's a config that makes sense.)

Reply:
Re-reading the JPEG spec was not part of my plan for today, but that was instructing.
So everything is good for this PR, however I don't think that you're right on the second part. But that's for the better!

> A.2 Order of source image data encoding
The scan header (see B.2.3) specifies the order by which source image data units shall be encoded and placed within the compressed image data. For a given scan, if the scan header parameter Ns = 1, then data from only one source component – the component specified by parameter Cs1 – shall be present within the scan. This data is non-interleaved by definition. If Ns > 1, then data from the Ns components Cs 1 through Cs Ns shall be present within the scan. This data shall always be interleaved. The order of components in a scan shall be according to the order specified in the frame header.

From that quote, I understand that you don't choose to interleave components or not. A single channel is not interleaved, while multi-channel are always interleaved.

> (We still get this wrong if there are sampling factors set and it's a multi-channel non-interleaved image I think. And that's a config that makes sense.)

Have you seen any image like that? Images like this should not exist if I understand the above correctly.

Reply:
> From that quote, I understand that you don't choose to interleave components or not. A single channel is not interleaved, while multi-channel are always interleaved.

[...]

> Have you seen any image like that? Images like this should not exist if I understand the above correctly.

Tests/LibGfx/test-inputs/jpg/several_scans_odd_number_mcu.jpg (https://github.com/SerenityOS/serenity/pull/17709#issuecomment-1915218787 :)) is almost like that. It has multiple components, but one per component. It even 4:2:0 subsampling, it just doesn't contain reset markers. But it could, from what I understand.

(I haven't seen this in the wild though.)

Reply:
Oh ok that's what you're calling "multi-channel non-interleaved", I only considered single scan images. Make sense now.
It just duplicates the scan header for no reason, so no decoder should generate them by default, but adding support for this kind of image was a good introduction for progressive JPEGs.